File: Star/Star.java
// test792d : USAG-Lib star

import java.io.*;
import java.nio.charset.StandardCharsets;

public class Star implements Closeable {
    // TarWriter
    private OutputStream out;

    // TarReader
    private InputStream in;
    private boolean isEOF;

    // Entry metadata
    public String name;
    public long size;
    public int mode;
    public boolean isDir;

    public Star() {
        this.out = null;
        this.in = null;
        this.isEOF = false;
        this.name = "";
        this.size = 0;
        this.mode = 0644;
        this.isDir = false;
    }

    // ========== Writer Methods ==========
    public void openWriter(OutputStream out) {
        if (out == null) { // memory stream
            this.out = new ByteArrayOutputStream();
        } else {
            this.out = out;
        }
    }

    public void write(String name, byte[] data, int mode) throws IOException {
        write(name, new ByteArrayInputStream(data), data.length, mode, false);
    }

    public void write(String name, File data, int mode) throws IOException {
        write(name, new FileInputStream(data), data.length(), mode, false);
    }

    public void write(String name, InputStream data, long size, int mode, boolean isDir) throws IOException {
        if (isDir) {
            name = name.replace('\\', '/');
            if (!name.endsWith("/")) name += "/";
        }
        if (name.getBytes(StandardCharsets.UTF_8).length > 99 || size > 077777777777L) { // needs PAX
            byte[] pax = paxHeader(name, size);
            this.out.write(pax);
        }
        this.out.write(tarHeader(name, size, mode, isDir ? '5' : '0'));
        if (!isDir) {
            copy(data, this.out, size);
            this.out.write(pad(size));
        }
    }

    private byte[] pad(long size) {
        int pad = (int) (512 - (size % 512));
        return new byte[pad % 512];
    }

    private byte[] tarHeader(String name, long size, int mode, char type) {
        byte[] h = new byte[512];
        if (size > 077777777777L) {
            size = 0;
        }

        // Name (0-100)
        writeStr(h, 0, name, 100);

        // Mode (100-108)
        writeOct(h, 100, mode, 8);

        // Size (124-136)
        writeOct(h, 124, size, 12);

        // MTime (136-148)
        writeOct(h, 136, System.currentTimeMillis() / 1000, 12);

        // Checksum (148-156) - Initially spaces
        for(int i=148; i<156; i++) h[i] = 32;

        // Typeflag (156)
        h[156] = (byte) type;

        // Magic (257) "ustar\0"
        writeStr(h, 257, "ustar", 6);

        // Version (263) "00"
        writeStr(h, 263, "00", 2);

        // Calculate Checksum
        long checksum = 0;
        for (byte b : h) checksum += (b & 0xFF); // unsigned sum
        String chkStr = String.format("%06o", checksum);
        System.arraycopy(chkStr.getBytes(StandardCharsets.UTF_8), 0, h, 148, 6);
        h[154] = 0;
        h[155] = 32;
        return h;
    }

    private byte[] paxHeader(String name, long size) {
        String result = "";
        String[] keys = {"path", "size"};
        String[] vals = {name, String.valueOf(size)};

        for(int i=0; i<2; i++) {
            String key = keys[i];
            String val = vals[i];
            String lineData = " " + key + "=" + val + "\n";
            
            // Calculate length iteratively
            int len = lineData.getBytes(StandardCharsets.UTF_8).length + 1;
            int currentLen = len;
            while(true) {
                String fullLine = currentLen + lineData;
                int actualLen = fullLine.getBytes(StandardCharsets.UTF_8).length;
                if (actualLen == currentLen) {
                    result += fullLine;
                    break;
                }
                currentLen = actualLen;
            }
        }
        
        // Return Header + Data + Padding
        byte[] paxData = result.getBytes(StandardCharsets.UTF_8);
        byte[] header = tarHeader("PaxHeader/" + name, paxData.length, 0644, 'x');
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        try {
            baos.write(header);
            baos.write(paxData);
            baos.write(pad(paxData.length));
        } catch (IOException e) { /* impossible */ }
        return baos.toByteArray();
    }

    // ========== Reader Methods ==========
    public void openReader(InputStream in) {
        this.in = in;
        this.isEOF = false;
        this.name = "";
        this.size = 0;
        this.mode = 0644;
        this.isDir = false;
    }

    private void unpad(long size) throws IOException {
        long pad = ((512 - (size % 512)) % 512);
        if (pad > 0) {
            byte[] temp = new byte[(int)pad];
            this.in.read(temp);
        }
    }

    private void parse(byte[] pax) {
        String data = new String(pax, StandardCharsets.UTF_8);
        String[] lines = data.split("\n");
        for (String line : lines) {
            int spaceIdx = line.indexOf(' ');
            if (spaceIdx == -1) continue;
            int eqIdx = line.indexOf('=');
            if (eqIdx == -1) continue;
            
            String key = line.substring(spaceIdx + 1, eqIdx);
            String value = line.substring(eqIdx + 1);
            if (key.equals("path")) this.name = value;
            else if (key.equals("size")) this.size = Long.parseLong(value);
        }
    }

    public boolean next() throws IOException {
        if (this.isEOF) return false;
        byte[] header = new byte[512];
        if (this.in.read(header, 0, 512) != 512) {
            this.isEOF = true;
            return false;
        }
        boolean isZero = true; // EOF check
        for (byte b : header) { if (b != 0) { isZero = false; break; } }
        if (isZero) {
            this.isEOF = true;
            return false;
        }

        // Parse Standard Header
        this.name = readStr(header, 0, 100).replace("\0", "");
        this.mode = (int) readOct(header, 100, 8);
        this.size = readOct(header, 124, 12);
        char type = (char) header[156];
        this.isDir = (type == '5');

        // Parse PAX
        if (type == 'x') {
            byte[] paxData = new byte[(int)this.size];
            this.in.read(paxData, 0, (int)this.size);
            unpad(this.size);
            parse(paxData);
            String paxName = this.name;
            long paxSize = this.size;

            boolean hasNext = next();
            this.name = paxName; // restore name
            this.size = paxSize;
            return hasNext;
        }
        return true;
    }

    public byte[] read() throws IOException {
        if (this.size > Integer.MAX_VALUE) throw new IOException("File too large for byte array");
        byte[] data = new byte[(int)this.size];
        int total = 0;
        while(total < this.size) {
            int r = this.in.read(data, total, (int)(this.size - total));
            if (r == -1) break;
            total += r;
        }
        unpad(this.size);
        return data;
    }

    public void mkfile(OutputStream dst) throws IOException {
        if (this.isDir) return;
        copy(this.in, dst, this.size);
        unpad(this.size);
    }

    public void skip() throws IOException {
        byte[] buffer = new byte[65536];
        long total = 0;
        while (total < this.size) {
            int toRead = (int)Math.min(buffer.length, this.size - total);
            int r = this.in.read(buffer, 0, toRead);
            if (r == -1) break;
            total += r;
        }
        unpad(this.size);
    }

    // ========== Common Methods ==========
    public byte[] closeTar() throws IOException {
        byte[] result = null;
        if (this.out != null) {
            this.out.write(new byte[1024]); // Two 512-byte blocks of zeros
            this.out.close();
            if (this.out instanceof ByteArrayOutputStream) {
                result = ((ByteArrayOutputStream)this.out).toByteArray();
            }
            this.out = null;
        }
        if (this.in != null) {
            this.in.close();
            this.in = null;
        }
        return result;
    }

    @Override public void close() throws IOException { closeTar(); }

    private void writeStr(byte[] b, int off, String s, int len) {
        byte[] strBytes = s.getBytes(StandardCharsets.UTF_8);
        System.arraycopy(strBytes, 0, b, off, Math.min(strBytes.length, len));
    }

    private void writeOct(byte[] b, int off, long v, int len) {
        String s = String.format("%0" + (len - 1) + "o", v); // octal format filled with 0s
        System.arraycopy(s.getBytes(StandardCharsets.UTF_8), 0, b, off, s.length());
    }

    private String readStr(byte[] b, int off, int len) {
        return new String(b, off, len, StandardCharsets.UTF_8);
    }

    private long readOct(byte[] b, int off, int len) {
        String s = new String(b, off, len, StandardCharsets.UTF_8).trim().replace("\0", "");
        return s.isEmpty() ? 0 : Long.parseLong(s, 8);
    }

    private void copy(InputStream in, OutputStream out, long size) throws IOException {
        byte[] buf = new byte[65536];
        int len;
        long total = 0;
        while (total < size && (len = in.read(buf, 0, (int)Math.min(buf.length, size-total))) != -1) {
            out.write(buf, 0, len);
            total += len;
        }
    }
}

File: Star/Star.js
// test792b : USAG-Lib star
// !!! JS version is not designed for big data !!!

const isNode = typeof process !== 'undefined' && process.versions != null && process.versions.node != null;
let fs;
if (isNode) {
    fs = require('fs');
}

class TarWriter {
    /**
     * @param {string} output set empty for memory, filepath(Node) or filename(Browser)
     */
    constructor(output) {
        this.outputStr = output;
        this.blocks = [];
        this.isMemory = (output === "");
        this.encoder = new TextEncoder();
    }

    /**
     * @param {string} name file name in tar
     * @param {string|Blob|File|Uint8Array} src file path (Node) or Blob/File object (Browser)
     * @param {boolean} isDir set true for directory
     */
    async write(name, src, isDir) {
        let data;
        
        // load data from src if not directory
        if (isDir) {
            name = name.replace(/\/?$/, '/');
            if (!name.endsWith('/')) name += '/';
            data = new Uint8Array(0);
        } else if (isNode) {
            if (typeof src === 'string') {
                data = await fs.promises.readFile(src);
            } else {
                data = src; 
            }
        } else {
            if (src instanceof Blob) {
                const arrayBuffer = await src.arrayBuffer();
                data = new Uint8Array(arrayBuffer);
            } else if (src instanceof Uint8Array) {
                data = src;
            } else {
                throw new Error("write in browser needs Blob or Uint8Array");
            }
        }
        
        if (isNode && Buffer.isBuffer(data)) {
            data = new Uint8Array(data);
        }
        this._add(name, data, isDir);
    }

    async close() {
        // Two 512-byte blocks of zeroes to mark end of archive
        this.blocks.push(new Uint8Array(512));
        this.blocks.push(new Uint8Array(512));

        let totalSize = 0; // join blocks
        for (const b of this.blocks) {
            totalSize += b.length;
        }
        const result = new Uint8Array(totalSize);
        let offset = 0;
        for (const b of this.blocks) {
            result.set(b, offset);
            offset += b.length;
        }
        this.blocks = [];

        if (this.isMemory) {
            return result;
        } else {
            if (isNode) {
                fs.writeFileSync(this.outputStr, result);
            } else {
                this._browserDownload(result, this.outputStr || "archive.tar");
            }
        }
        return null;
    }

    _add(name, data, isDir) {
        let needPax = false; // PAX condition check
        if (this.encoder.encode(name).length > 99 || data.length > 0o77777777777) {
            needPax = true;
        }

        if (needPax) { // Write PAX Header if needed
            let pax = this._createPax(name, data.length)
            this.blocks.push(this._createHeader("PaxHeader/" + name, pax.length, 'x'));
            this.blocks.push(pax);
            let pad = this._createPad(pax.length);
            if (pad != null) {
                this.blocks.push(pad);
            }
        }
        this.blocks.push(this._createHeader(name, data.length, isDir ? '5' : '0')); // Write USTAR Header
        if (!isDir) {
            this.blocks.push(data); // Write File Data
            let pad = this._createPad(data.length);
            if (pad != null) {
                this.blocks.push(pad);
            }
        }
    }

    _createPad(size) {
        const padSize = (512 - (size % 512)) % 512;
        let data = null;
        if (padSize > 0) {
            data = new Uint8Array(padSize);
        }
        return data;
    }

    _createPax(name, size) {
        let records = ""; // Format: "length keyword=value\n"
        let data = [["path", name], ["size", size.toString()]];
        for (const pair of data) {
            const key = pair[0];
            const value = pair[1];
            let lenStr = "0"; 
            let totalLen = 0;
            
            while (true) { // retry until length is valid
                totalLen = lenStr.length + 1 + key.length + 1 + value.length + 1;
                const newLenStr = totalLen.toString();
                if (newLenStr.length === lenStr.length) {
                    lenStr = newLenStr;
                    break;
                }
                lenStr = newLenStr; // retry with new length
            }
            records += `${lenStr} ${key}=${value}\n`;
        }
        return this.encoder.encode(records);
    }

    _createHeader(name, size, typeflag) {
        const header = new Uint8Array(512);
        header.fill(0);
        const mtime = Math.floor(Date.now() / 1000);
        if (size > 0o77777777777) {
            size = 0 // Will be set by PAX
        }
        
        // Name (0-100)
        this._writeString(header, 0, name, 100);

        // Mode (100-108) - 0644
        this._writeOctal(header, 100, 0o644, 8);

        // Size (124-136)
        this._writeOctal(header, 124, size, 12);

        // MTime (136-148)
        this._writeOctal(header, 136, mtime, 12);

        // Checksum (148-156) - calculated later
        for(let i=148; i<156; i++) header[i] = 32; // Spaces

        // Typeflag (156)
        header[156] = typeflag.charCodeAt(0);

        // Magic (257-263) - "ustar"
        this._writeString(header, 257, "ustar", 6);
        
        // Version (263-265) - "00"
        this._writeString(header, 263, "00", 2);

        // Calculate Checksum
        let sum = 0;
        for (let i = 0; i < 512; i++) sum += header[i];
        const chksumStr = sum.toString(8).padStart(6, '0');
        for(let i=0; i<6; i++) header[148+i] = chksumStr.charCodeAt(i);
        header[154] = 0;
        header[155] = 32;
        return header;
    }

    _writeString(buf, offset, str, len) {
        const encoded = this.encoder.encode(str);
        for (let i = 0; i < len; i++) {
            buf[offset + i] = (i < encoded.length) ? encoded[i] : 0;
        }
    }

    _writeOctal(buf, offset, num, len) {
        const str = num.toString(8).padStart(len - 1, '0');
        for (let i = 0; i < len - 1; i++) {
            buf[offset + i] = str.charCodeAt(i);
        }
        buf[offset + len - 1] = 0;
    }

    _browserDownload(data, filename) {
        const blob = new Blob([data], { type: "application/octet-stream" });
        const url = URL.createObjectURL(blob);
        const a = document.createElement("a");
        a.style.display = "none";
        a.href = url;
        a.download = filename;
        document.body.appendChild(a);
        a.click();
        
        setTimeout(() => {
            document.body.removeChild(a);
            window.URL.revokeObjectURL(url);
        }, 100);
    }
}

class TarReader {
    /**
     * @param {string|Blob|Uint8Array} input set empty for memory, filepath(Node) or filename(Browser)
     */
    constructor(input) {
        this.input = input;
        this.data = null;
        this.decoder = new TextDecoder();
        this.files = null; // {name, size, offset, isDir}
    }

    async init() {
        // load data
        if (isNode && typeof this.input === 'string') {
            this.data = await fs.promises.readFile(this.input);
        } else if (typeof this.input === 'object' && this.input instanceof Blob) {
            const buffer = await this.input.arrayBuffer();
            this.data = new Uint8Array(buffer);
        } else if (this.input instanceof Uint8Array) {
            this.data = this.input;
        } else {
            throw new Error("Invalid input type");
        }
        let current = 0;
        this.files = [];

        // metadata
        let wasPax = false;
        let name = "";
        let size = 0;
        let isDir = false;

        while (current + 512 <= this.data.length) {
            let isZero = true;
            for(let i=0; i<512; i++) {
                if (this.data[current + i] !== 0) {
                    isZero = false;
                    break;
                }
            }
            if (isZero) break; // End of archive

            // name, size info
            let tsize = this._readOctal(this.data.subarray(current + 124, current + 136));
            if (!wasPax) {
                name = this._readString(this.data.subarray(current, current + 100));
                size = tsize;
            }

            // typeflag info
            const typeflag = String.fromCharCode(this.data[current + 156]);
            if (typeflag === 'x') { // PAX header
                const paxData = this.data.subarray(current + 512, current + 512 + tsize);
                const pax = this._parsePax(paxData);
                name = pax.name;
                size = pax.size;
                current += 512 + Math.ceil(tsize / 512) * 512;
                wasPax = true;
                continue; // Skip to next header
            } else if (typeflag === '5') { // Directory
                isDir = true;
            }
            this.files.push({
                name: name,
                size: size,
                offset: current + 512,
                isDir: isDir
            });

            // Move to next header
            current += 512 + Math.ceil(size / 512) * 512;
            wasPax = false;
            name = "";
            size = 0;
            isDir = false;
        }
    }

    read(idx) {
        if (idx < 0 || idx >= this.files.length) throw new Error("Index out of bounds");
        const file = this.files[idx];
        return this.data.subarray(file.offset, file.offset + file.size);
    }

    close() {
        this.input = null;
        this.data = null;
        this.files = [];
    }

    _readString(bytes) {
        let end = bytes.indexOf(0);
        if (end === -1) end = bytes.length;
        return this.decoder.decode(bytes.subarray(0, end));
    }

    _readOctal(bytes) {
        let str = this.decoder.decode(bytes);
        str = str.replace(/\0/g, '').trim(); 
        return parseInt(str, 8) || 0;
    }

    _parsePax(data) {
        const str = this.decoder.decode(data); // Format: length keyword=value\n
        let path = "";
        let size = 0;
        let pos = 0;
        while (pos < str.length) {
            const spaceIdx = str.indexOf(' ', pos);
            if (spaceIdx === -1) break;
            const eqIdx = str.indexOf('=', spaceIdx);
            if (eqIdx === -1) break;
            const lfIdx = str.indexOf('\n', eqIdx);
            if (lfIdx === -1) break;
            pos = lfIdx + 1;

            const key = str.substring(spaceIdx + 1, eqIdx);
            const value = str.substring(eqIdx + 1, lfIdx);
            if (key === 'path') path = value;
            else if (key === 'size') size = parseInt(value, 10);
        }
        return { name: path, size: size };
    }
}